homeassistant/components/ozw/migration.py

```diff
@@ -0,0 +1,169 @@
+            if unique_id not in unique_entries:
+                continue
+            node = entity_values.primary.node
+            device_identifier = (
+                DOMAIN,
+                create_device_id(node, entity_values.primary.instance),
+            )
+            device_entry = dev_reg.async_get_device({device_identifier}, set())
+            data[unique_id] = {
+                "node_id": node_id,
+                "node_instance": entity_values.primary.instance,
+                "device_id": device_entry.id,
+                "command_class": entity_values.primary.command_class.value,
+                "command_class_label": entity_values.primary.label,
+                "value_index": entity_values.primary.index.value,
+                "unique_id": unique_id,
+                "entity_entry": unique_entries[unique_id],
+            }
+
+    return data
+
+
+def map_node_values(zwave_data, ozw_data):
+    """Map zwave node values onto ozw node values."""
+    migration_map = {"device_entries": {}, "entity_entries": {}}
+
+    for zwave_entry in zwave_data.values():
+        node_id = zwave_entry["node_id"]
+        node_instance = zwave_entry["node_instance"]
+        cc_id = zwave_entry["command_class"]
+        zwave_cc_label = zwave_entry["command_class_label"]
+
+        if cc_id in CC_ID_LABELS:
+            labels = CC_ID_LABELS[cc_id]
+            ozw_cc_label = labels.get(zwave_cc_label, zwave_cc_label)
+
+            ozw_entry = next(
+                (
+                    entry
+                    for entry in ozw_data.values()
+                    if entry["node_id"] == node_id
+                    and entry["node_instance"] == node_instance
+                    and entry["command_class"] == cc_id
+                    and entry["command_class_label"] == ozw_cc_label
+                ),
+                None,
+            )
+        else:
+            value_index = zwave_entry["value_index"]
+
+            ozw_entry = next(
+                (
+                    entry
+                    for entry in ozw_data.values()
+                    if entry["node_id"] == node_id
+                    and entry["node_instance"] == node_instance
+                    and entry["command_class"] == cc_id
+                    and entry["value_index"] == value_index
+                ),
+                None,
+            )
+
+        if ozw_entry is None:
+            continue
+
+        # Save the zwave_entry under the ozw entity_id to create the map.
+        migration_map["entity_entries"][
+            ozw_entry["entity_entry"].entity_id
+        ] = zwave_entry
+        migration_map["device_entries"][ozw_entry["device_id"]] = zwave_entry[
+            "device_id"
+        ]
+
+    return migration_map
+
+
+async def async_migrate(hass, migration_map):
+    """Perform zwave to ozw migration."""
+    dev_reg = await async_get_device_registry(hass)
+    for ozw_device_id, zwave_device_id in migration_map["device_entries"].items():
```

Comment:
Why are we migrating info between devices instead of merging them ? A device can be linked to multiple integrations: https://developers.home-assistant.io/docs/device_registry_index

Calling `dev_reg.async_get_or_create` with identifiers of both devices will link them up. We find by identifier first, so you just call `dev_reg.async_get_or_create` with the Z-Wave device ID first, so it gets OZW one added. Then delete OZW device ID.

Reply:
(this works too, so fine with it)

Reply:
We'd have to look up the device identifiers first to do that, right? So it looks like the current way is easier.

Reply:
Yeah, you will need to find both entries first anyway.

Reply:
We should apply this suggestion to solve the migration of device actions etc.

Reply:
We've decided to not add more to this migration, consider the recent move with zwave_js.